hibernate-core/src/main/java/org/hibernate/proxy/AbstractLazyInitializer.java

```diff
@@ -67,6 +67,8 @@ protected AbstractLazyInitializer(String entityName, Serializable id, SharedSess
 		}
 		else {
 			setSession( session );
+			initializeProxyWhenAccessingIdentifier = session.getFactory().getSessionFactoryOptions()
+					.getJpaCompliance().isJpaProxyComplianceEnabled();
 		}
```

Comment:
I don't think it's enough for ```initializeProxyWhenAccessingIdentifier``` to be a field.

Here is a scenario where there could be problems:
1) the object is constructed, using a ```SessionFactory``` with ```hibernate.jpa.compliance.proxy=false``` configured;
2) the object is detached and ```session`` is set to null;
3) the object is attached to a ```Session``` for a different ```SessionFactory``` with ```hibernate.jpa.compliance.proxy=true``` configured;
4) the application calls ```getIdentifier()```, which returns the ID without initializing the entity.

Since ```hibernate.jpa.compliance.proxy=true``` is configured in the new ```SessionFactory```, the entity should have been initialized before returning the ID.

I think that the ```sessionFactoryUuid``` should also be assigned, and the ```SessionFactory``` should be checked in ```getIdentifier()```, similar to what is done for ```allowLoadOutsideTransaction```.

IMO, when the object gets reattached to a ```Session```, ```sessionFactoryUuid``` should be updated to the UUID for ```Session```'s ```SessionFactory```.